### ISSUE TITLE
Timing best_effort desynchronizes over time

### ISSUE THREAD

**Issue by pspacek** [DNS-OARC/drool#98]
Hello,

it seems that `timing best_effort` is somewhat broken.

I have a PCAP with about 2500 queries/second, which well within capabilities of machine I'm testing on.

When I run drool using this command:
```
drool -r ~/samples/odvr-1h-noany.pcap -c 'text:context client_pools 5000; client_pool sendas udp; client_pool target "127.0.0.1" "5353"; client_pool skip_reply; timing ignore;
```
it spits off a lot of warnings like this:

```
2018-02-12 15:39:34 callback.c:0245 network warning: Unable to keep up with timings (process cost 0.000006556, packet diff 0.000003, now 356997.108642455, sleep to 356997.108632343)
```

This is weird, because when I use `timing ignore`, the drool is able to replay the same PCAP in 10 seconds.

What is even weirder, if I replay the same PCAP using `timing best_effort` many times and measure number of packets replayed every second, then I can observe very small difference at the beginning but it gets bigger and bigger over time, sugesting Drool is getting more and mode desynchronized. I will post a graph to show this.

**Comment by pspacek** [comment]
Graph from packet number measurement:
[pktstat-10-sec-average.pdf](https://github.com/DNS-OARC/drool/files/1716680/pktstat-10-sec-average.pdf)
Points in the graph show average number of packets in 10 seconds measurement window. You can see that at the beginning lines from all three attempts exactly line up, but as the time progresses, the lines diverge. This indicates that Drool is not replaying packets in the same manner.

**Comment by jelu** [comment]
Why are you running 5000 client pools? That would start >5000 threads.

The command you posted has `timing ignore`, are you sure it's the right command?

`best_effort` is really just `keep` without any output when it can't.

I am not at all surprised that it may get desynched as further it runs, the timing code is in need of a rework #75 and this will get hard to get right since there is no absolute values coming from the PCAPs.

The more you measure the time took the more time it takes to measure and the harder it gets to get back in sync.

**Comment by pspacek** [comment]
Oh, copy&paste error from a looong terminal history with wrapped lines!

I've tried `timing best_effort` along with 500 and 5000 client_pools, withhout a significant result. This was an attempt to see what happens if it helps to have more clients than seen in PCAP, but without result.

Speaking of PCAP, I'm confused ... I always thought that there is absolute timestamp in it, and that is also my interpretation of https://wiki.wireshark.org/Development/LibpcapFileFormat#Record_.28Packet.29_Header . Is there a misunderstanding?

What I woud like to see is to schedule packets as precisely as possible, as when they are late for some reason, schedule at least the packets after them to the precise time (derived from the PCAP).

**Comment by jelu** [comment]
`context client_pools` are not clients, it's pools of clients. Use `client_pool max_clients NUMBER ;` instead. For good performance reading one PCAP use number of CPUs - 1 for `client_pools`.

Yeah your right, timestamps are absolute, my bad. But that does not change that the timing code needs a good rework/rethink.

You are more then happy to go ahead and try to take a stab at the timing code :)

I consider this issues a duplicate of #75.

**Comment by pspacek** [comment]
Okay, I will give it a try with new parameters. Thank you for advice!

**Comment by jelu** [comment]
@pspacek some questions:
- There are 3 measurements in the graph, what are they from?
- I assume the graph shows QPS on Y axis and seconds on X axis, correct?
- Did you use a lot of `-v` when measuring?
- Is there TCP queries in the PCAP replayed?

**Comment by pspacek** [comment]
For the record:
`-v` was not used during measurement so logging should not slow things down.

To graphs:
Yes, these curves show QPS on the interface. I've captured traffic using `tcpdump` into file and then produced stats using `tracertstats` on that file. The three lines on graph come from three distinct runs with the same parameters.

Last time I've measured it using `text:context client_pools 10; client_pool max_clients 500; client_pool sendas udp; client_pool target "127.0.0.1" "5353"; client_pool skip_reply; timing best_effort;` but it didn't make any significant difference.

It turns out that my measurement was significantly skewed by incorrect `tcpdump` filter so it traffic produced by Drool was mixed with reply traffic.

With correct filter (`dst port 5353`) the result it much better, but not perfect. See new graph:
[drool-timing.pdf](https://github.com/DNS-OARC/drool/files/1724408/drool-timing.pdf)
There are 10 distinct curves in this graph, the these should exactly match up - but they don't.

I'm going to provide proof-of-concept PR which improves this.
